### PR TITLE
Reconfigure 'select all' for texts in macOS.

### PR DESCRIPTION
**Pull request by galegosimpatico**
Use the idiomatic combination.

### CODE REVIEW

**Comment by CelticMinstrel** [comment]
Somehow I never noticed this one. I do think it's silly to have two variables that are basically the same thing, though - I would've preferred renaming `copypaste_modifier` to be less specific and the just using it also for select all.

**Comment by galegosimpatico** [comment]
I think this would be important, in a path towards a macOS fully idiomatic (and user configuration sentient if/when possible) _save game box_.

I would never noticed as well. In my case, these kind of things are only realized when I am in it and the _gotcha_ just pops out unexpectedly breaking the principle of least suprise.

You are right about your suggestion. I will be analyzing in more depth the comment on `copypaste_modifier`, find the referenced comment to understand the whole comment including reference, and trying to simplify it or have it completely removed. Finally melting both variables in a single.

Thanks and cheers and regards.

File: src/gui/widgets/text_box_base.cpp [hunk 1]
```diff
@@ -478,13 +478,28 @@ void text_box_base::signal_handler_sdl_key_down(const event::ui_event event,
...
+// aligned in a same direction, i.e. macOS defaults.
+
+// The election for leader key is not pervasive and, even though configurable
+// (or at least it _should_ be configurable), is very often set to what the
+// desktop environment defines as idiomatic.
+
+#ifndef __APPLE__
+	// Idiomatic leader key in GNU/Linux desktop environments and
+	// Microsoft desktop environments.
+	const unsigned leader_key = KMOD_CTRL;
```
Comment (Vultraz) [inline src/gui/widgets/text_box_base.cpp]: I don't think this should be called `leader_key`. Something like `modifier_key` would be better.

File: src/gui/widgets/text_box_base.cpp [hunk 2]
```diff
@@ -478,13 +478,28 @@ void text_box_base::signal_handler_sdl_key_down(const event::ui_event event,
...
+#else // __APPLE__
+	// Idiomatic leader key in macOS computers.
+	const unsigned leader_key = KMOD_GUI;
 #endif
 
+	// No one still uses BSD don't they?
+
+	const auto copypaste_modifier = leader_key;
+	const auto select_all_leader_key = leader_key;
```
Comment (Vultraz) [inline src/gui/widgets/text_box_base.cpp]: Unnecessary extra variable.

File: src/gui/widgets/text_box_base.cpp [hunk 3]
```diff
@@ -478,13 +478,28 @@ void text_box_base::signal_handler_sdl_key_down(const event::ui_event event,
...
+	// Microsoft desktop environments.
+	const unsigned leader_key = KMOD_CTRL;
+#else // __APPLE__
+	// Idiomatic leader key in macOS computers.
+	const unsigned leader_key = KMOD_GUI;
 #endif
 
+	// No one still uses BSD don't they?
+
+	const auto copypaste_modifier = leader_key;
```
Comment (Vultraz) [inline src/gui/widgets/text_box_base.cpp]: Unnecessary extra variable. Even if you wanted to keep these, you should be using `unsigned` instead of `auto`.

**Comment by Vultraz** [comment]
Why do you need two variables with the same value?

**Comment by galegosimpatico** [comment]
Now I was feeling most those constants should become variables actually, and configurable, eventually.

**Comment by galegosimpatico** [comment]
I mean, it's the best way to reconcile these:

* Minimize gotchas experienced by macOS enabled newcomers. _Being nice to the new_.

* Stay consistent with the spirit of fully configuration shining there in `preferences > hotkeys`. _Being loyal to the old_.

**Comment by Vultraz** [comment]
Alright, if you think it’s better to have two variables, that’s fine, but you should document why both exist.

**Comment by CelticMinstrel** [comment]
These shortcuts aren't configurable and I see little reason why they should become so (though I wouldn't be opposed to it).

File: src/gui/widgets/text_box_base.cpp [hunk 4]
```diff
@@ -478,11 +478,27 @@ void text_box_base::signal_handler_sdl_key_down(const event::ui_event event,
...
+
 #ifdef __APPLE__
-	const unsigned copypaste_modifier = KMOD_LGUI | KMOD_RGUI;
+	// Idiomatic modifier keys in macOS computers for some _gestures_.
+	const SDL_Keycode copypaste_modifier = KMOD_GUI;
+	const SDL_Keycode select_all_modifier = KMOD_GUI;
 #else
-	const unsigned copypaste_modifier = KMOD_CTRL;
+	// Idiomatic modifier keys in Microsoft desktop environments for some
+	// _gestures_. _Common_ in GNU/Linux desktop environments as well.
```
Comment (CelticMinstrel) [inline src/gui/widgets/text_box_base.cpp]: Why the extra underscores in here? I guess it doesn't matter _that_ much, but it makes it look like you're treating the comment content as Markdown, which doesn't make any sense.

**Comment by galegosimpatico** [comment]
If they become configurable in the future that's the very reason to have it separately... to allow (even encourage it a little bit) to be remapped progressively.

I don't even know how likely are each them to become remappable nor which one could become a more popular remapping.

Maybe I got the source wrong, or may I be missing large parts of the general picture (you can be sure), but it seems to me that is precisely _configuring each shortcut with the modifier key of choice_ what is reason to have them as separate `const`s, I don't even know if each should cease to be `const` whenever becoming remappable, or if each would need different treatment.

File: src/gui/widgets/text_box_base.cpp (hunk 4 above)
Comment (galegosimpatico) [reply src/gui/widgets/text_box_base.cpp]: Wouldn't have made any sense unless Markdown had been conceived specifically to provide source readability... but that's the way they designed it.

**Comment by CelticMinstrel** [comment]
No, my point is if they became remappable in the future, _both_ variables become irrelevant. This entire function would look totally different. You'd be able to assign Shift+Insert to paste and Ctrl+Insert to copy, if you wanted. We wouldn't just add a preference letting people choose the modifier to use for these keys; we'd let them choose the exact keystroke sequence to use.

**Comment by Vultraz** [comment]
We aren’t making these configurable.

**Comment by galegosimpatico** [comment]
OK. I was expecting that the _title screen_ column in `preferences > hotkeys`...

screen shot 2017-10-17 at 20 11 16

... could eventually hold these _misc gui_ hotkeys, hotkey combinations, whatever; or they even could eventually earn a column on their own. I haven't followed the evolution of this, have had the editor opening **once** ever, so know all those details too when pondering this opinion.

These last hours my understanding has been that all four _copy_, _cut_, _paste_ and _select all_ involved in this PR are _editor_ `Actions` as well as _title screen / misc gui_ `Actions`.

So... this is countering, but the intention is only me making sure I don't grow on you any second thoughts on these as happened to me. I can let you some days just in case you change your mind and then I am reverting to single constant.

> You'd be able to assign Shift+Insert to paste and Ctrl+Insert to copy, if you wanted. We wouldn't just add a preference letting people choose the modifier to use for these keys; we'd let them choose the exact keystroke sequence to use.

Perfect, this reads as you won't change your mind. That's like the project-familiar point of view I was needing.

I was just trying to be over cautious.

Thank you!

**Comment by CelticMinstrel** [comment]
> We aren’t making these configurable.

I don't think there's any reason to utterly rule it out.

**Comment by galegosimpatico** [comment]
I'll be proposing other minor `dev` changes under `Preferences > Hotkeys` anyway.

Really unlikely those can be ready before `1.14.0` makes it to GA.

Not willing to disclose anything else but-and-because I'm very confident you will embrace.

**Comment by Vultraz** [comment]
Looks good to me now.

**Comment by galegosimpatico** [comment]
> > We aren’t making these configurable.
> > 
> I don't think there's any reason to utterly rule it out.
> 
I remembered, in my country many people would prefer using `e` to _select all_, instead of `a`. Not sure about those using macOS because I haven't used it localized, but surely that would be idiomatic to many people using Microsoft.

Those people would be expecting some hotkey combinations to be defaulting depending not only on device vendor but also on locale.

While sources for that have not been contributed, I don't see the point in discussing it now.

**Pull request merged**